File: medium/790. Domino and Tromino Tiling/790. Domino and Tromino Tiling.py
class Solution:
    def numTilings(self, n: int) -> int:
        return sol2(n)


def sol1(n: int) -> int:
    F, T, B = {0: 1, 1: 1}, {1: 0}, {1: 0}

    for i in range(2, n + 1):
        F[i] = F[i - 1] + F[i - 2] + T[i - 1] + B[i - 1]
        T[i] = B[i - 1] + F[i - 2]
        B[i] = F[i - 2] + T[i - 1]

    return F[n] % (10**9 + 7)


def sol2(n: int) -> int:
    FF, F, T, B = 1, 1, 0, 0

    for i in range(2, n + 1):
        FF, F, T, B = F, F + FF + B + T, B + FF, T + FF

    return F % (10**9 + 7)

File: medium/790. Domino and Tromino Tiling/790. Domino and Tromino Tiling.go
package main

const MOD = 1_000_000_007

func numTilings(n int) int {
	FF, F, T, B := 1, 1, 0, 0

	for i := 2; i <= n; i++ {
		FF, F, T, B = F, (F+FF+T+B)%MOD, (FF+B)%MOD, (FF+T)%MOD
	}

	return F
}
